src-cpp/rdkafkacpp.h

```diff
@@ -2927,50 +3031,150 @@ class RD_EXPORT Producer : public virtual Handle {
+   *          or fatal error has been raised by calling
+   *          RdKafka::Error::is_txn_abortable() or RdKafka::Error::is_fatal()
+   *          respectively.
+   *
+   * @remark The returned error object (if not NULL) must be deleted.
+   *
+   * See rd_kafka_send_offsets_to_transaction() in rdkafka.h for
+   * more information.
    */
-  virtual ErrorCode send_offsets_to_transaction (
+  virtual Error *send_offsets_to_transaction (
           const std::vector<TopicPartition*> &offsets,
-          const std::string &group_id,
-          int timeout_ms,
-          std::string &errstr) = 0;
+          const ConsumerGroupMetadata *group_metadata,
+          int timeout_ms) = 0;
 
   /**
-   * @brief
+   * @brief Commit the current transaction (as started with begin_transaction()).
+   *
+   *        Any outstanding messages will be flushed (delivered) before actually
+   *        committing the transaction.
+   *
+   * @param timeout_ms The maximum time to block. On timeout the operation
+   *                   may continue in the background, depending on state,
+   *                   and it is okay to call this function again.
    *
-   * FIXME blocking?
+   * @returns an RdKafka::Error object on error, or NULL on success.
+   *          Check whether the returned error object permits retrying
+   *          by calling RdKafka::Error::is_retriable(), or whether an abortable
+   *          or fatal error has been raised by calling
+   *          RdKafka::Error::is_txn_abortable() or RdKafka::Error::is_fatal()
+   *          respectively.
+   *
+   * @remark The returned error object (if not NULL) must be deleted.
+   *
+   * See rd_kafka_commit_transaction() in rdkafka.h for more information.
    */
-  virtual ErrorCode commit_transaction (int timeout_ms,
-                                        std::string &errstr) = 0;
+  virtual Error *commit_transaction (int timeout_ms) = 0;
 
   /**
-   * @brief
+   * @brief Aborts the ongoing transaction.
+   *
+   *        This function should also be used to recover from non-fatal abortable
+   *        transaction errors.
+   *
+   *        Any outstanding messages will be purged and fail with
+   *        RdKafka::ERR__PURGE_INFLIGHT or RdKafka::ERR__PURGE_QUEUE.
+   *        See RdKafka::Producer::purge() for details.
+   *
+   * @param timeout_ms The maximum time to block. On timeout the operation
+   *                   may continue in the background, depending on state,
+   *                   and it is okay to call this function again.
    *
-   * FIXME blocking?
+   * @returns an RdKafka::Error object on error, or NULL on success.
+   *          Check whether the returned error object permits retrying
+   *          by calling RdKafka::Error::is_retriable(), or whether an abortable
+   *          or fatal error has been raised by calling
+   *          RdKafka::Error::is_txn_abortable() or RdKafka::Error::is_fatal()
```

Comment:
what does it mean to have `is_txn_abortable` set on an error returned by `rd_kafka_abort_transaction`? if this can indeed be set here, I think it probably needs some additional commentary.

Reply:
You're right, abort_transaction() should not return an txn_abortable() error, but rather retriable().